homeassistant/components/zha/entities/listeners.py

```diff
@@ -0,0 +1,71 @@
+"""
+Cluster listeners for Zigbee Home Automation.
+
+For more details about this component, please refer to the documentation at
+https://home-assistant.io/components/zha/
+"""
+
+
+class ClusterListener:
+    """Listener for a Zigbee cluster."""
+
+    def __init__(self, entity):
+        """Initialize OnOffListener."""
+        self._entity = entity
+
+    def cluster_command(self, tsn, command_id, args):
+        """Handle commands received to this cluster."""
+        pass
+
+    def attribute_updated(self, attrid, value):
+        """Handle attribute updates on this cluster."""
+        pass
+
+    def zdo_command(self, *args, **kwargs):
+        """Handle ZDO commands on this cluster."""
+        pass
+
+    def zha_send_event(self, cluster, command, args):
+        """Relay entity events to hass."""
+        pass  # don't let entities fire events
+
+
+class OnOffListener(ClusterListener):
+    """Listener for the OnOff Zigbee cluster."""
+
+    def cluster_command(self, tsn, command_id, args):
+        """Handle commands received to this cluster."""
+        if command_id in (0x0000, 0x0040):
+            self._entity.set_state(False)
+        elif command_id in (0x0001, 0x0041, 0x0042):
+            self._entity.set_state(True)
+        elif command_id == 0x0002:
+            self._entity.set_state(not self._entity.is_on)
+
+    def attribute_updated(self, attrid, value):
+        """Handle attribute updates on this cluster."""
+        if attrid == 0:
+            self._entity.set_state(value)
+
+
+class LevelListener(ClusterListener):
+    """Listener for the LevelControl Zigbee cluster."""
+
+    def cluster_command(self, tsn, command_id, args):
+        """Handle commands received to this cluster."""
+        if command_id in (0x0000, 0x0004):  # move_to_level, -with_on_off
```

Comment:
replace magic #'s with constants

Reply:
I love debug logging. What if we do something like:
```
        cmd = self.cluster.server_commands.get(command_id, [command_id])[0]
        _LOGGER.debug(
            "%s: received '%s' command with %s args on 'Level' cluster",
            self._entity.entity_id, cmd, args)
         if cmd in ('move_to_level', 'move_to_level_with_on_off'):
```
I get logging and code is more descriptive?

Reply:
copy'n'pasted from another branch, but just wanted to relay the idea

Reply:
will do

Reply:
done